design/sticky-session-design.md

```diff
@@ -0,0 +1,170 @@
+
+## Goals
+
+- Enabling support for sticky session in Contour using
+    - Cookie
+    - Source IP
+    - Header
+- Adding support in IngressRoute CRD. 
+
+## Non-goals
+
+- Making options in IngressRoute very specific to Envoy is not a goal. Options in IngressRoute will be generic.
+
+## Background
+
+Sticky-sessions allow requests to be routed to the same host based on certain parameters of the request.
+The parameters can be a cookie, the source IP, a custom header, etc.
+
+Envoy supports sticky sessions by using a hash-based load balancing strategy (RingHash, Maglev). 
+Specifying a *hash policy* for such load-balancers will enable sticky sessions. However, we are not aiming at making
+this very specific to Envoy. IngressRoute options should be generic enough without exposing
+internals of Envoy itself.
+
+
+## High-Level Design
+
+Sticky sessions are enabled per route. The following change are required in IngressRoute:
+
+```yaml
+apiVersion: contour.heptio.com/v1beta1
+kind: IngressRoute
+metadata: 
+  name: basic
+  namespace: default
+spec: 
+  virtualhost:
+    fqdn: foo-basic.bar.com
+  routes: 
+    - match: /
+      sticky-session:
+        source: cookie/header/connection # There should be exactly one of these set.
+        name: name # This can be cookie name, header name, or connection property
+        ttl: time-to-live # This is only valid when source = cookie. 
+        path: cookie-path # This is only valid when source = cookie. 
+      services: 
+        - name: s1
+          port: 80  
+          strategy: RingHash # Optional. Default: RingHash
+```          
+
+For sticky sessions to be enabled on any route, the service associated to the route needs to have the 
+load balancing strategy set to a hash-based strategy (like RingHash or Maglev). The following cases need to be considered:
+
+- No service strategy is specified: In this case, contour will set the service's strategy to "RingHash" by default.
+- Service strategy specified, but not set to a Hash-based strategy: In this case, there will be no effect of declaring sticky session.
+
+
+## Detailed Design
+
+### IngressRoute
+- The following changes will be made in *ingressroute.go*:   
+```go
+
+type Route struct {
+	...
+	// Indicates that sticky session will be configured for the route
+	StickySession StickySessionConfig
+}
+
+
+// StickySession
+type StickySessionConfig struct {
+	Source string `json:"source"`
+	Name   string `json:"name"`
+	TTL    string `json:"ttl"`
+	Path   string `json:"path"`
+}
+```
+
+- deployment/common/common.yaml will be modified to include change in the ingressroute CRD.
+
+### DAG
+
+- The Route structure in dag.go will be changed to add the sticky session config
+```go
+type Route struct {
+	...
+	// Indicates that during forwarding, the matched prefix (or path) should be swapped with this value
+	PrefixRewrite string
+
+	// Stores Sticky session Config
+	StickySession ingressroutev1.StickySessionConfig
```

Comment:
I'd prefer it if the k8s types didn't leak into the dag. I know I haven't been scrupulous in avoiding this, but it should be a goal that the dag is built from k8s types, not built out of k8s types.

Reply:
Sure, I understand the motive of not using k8s types in the dag. I will change this.